File: Assignment6.java
import java.io.*;
import java.util.*;
import java.util.concurrent.locks.*;
import java.util.logging.*;

public class Assignment6 {
    private static final Logger logger = Logger.getLogger("RideShareLogger");
    private final Queue<String> taskQueue = new LinkedList<>();
    private final List<String> results = new ArrayList<>();
    private final Lock queueLock = new ReentrantLock();
    private final Lock resultsLock = new ReentrantLock();
    private final Condition notEmpty = queueLock.newCondition();

    public void addTask(String task) {
        queueLock.lock();
        try {
            taskQueue.add(task);
            notEmpty.signal();
        } finally {
            queueLock.unlock();
        }
    }

    public String getTask() throws InterruptedException {
        queueLock.lock();
        try {
            while (taskQueue.isEmpty()) {
                notEmpty.await();
            }
            return taskQueue.poll();
        } finally {
            queueLock.unlock();
        }
    }

    public void worker(int id) {
        Thread.currentThread().setName("Worker-" + id);
        logger.info(Thread.currentThread().getName() + " started.");
        while (true) {
            String task;
            try {
                task = getTask();
            } catch (InterruptedException ie) {
                logger.warning(Thread.currentThread().getName() + " interrupted. Exiting.");
                break;
            }
            try {
                Thread.sleep(new Random().nextInt(500) + 500); // simulate compute
                String result = Thread.currentThread().getName() + " processed: " + task;
                resultsLock.lock();
                try {
                    results.add(result);
                } finally {
                    resultsLock.unlock();
                }
                logger.info(result);
            } catch (InterruptedException ie) {
                logger.warning(Thread.currentThread().getName() + " sleep interrupted.");
                break;
            } catch (Exception e) {
                logger.log(Level.SEVERE, Thread.currentThread().getName() + " error: " + e.getMessage(), e);
            }
        }
        logger.info(Thread.currentThread().getName() + " exiting.");
    }

    public void run(int numWorkers) throws InterruptedException, IOException {
        // Setup logger to file:
        FileHandler fh = new FileHandler("rideshare.log");
        fh.setFormatter(new SimpleFormatter());
        logger.addHandler(fh);

        // Launch workers
        List<Thread> threads = new ArrayList<>();
        for (int i = 0; i < numWorkers; i++) {
            final int workerId = i;
            Thread t = new Thread(() -> worker(workerId));
            threads.add(t);
            t.start();
        }

        // Add sample tasks
        for (int i = 1; i <= 20; i++) {
            addTask("RideRequest#" + i);
        }

        // Wait for queue to empty
        while (true) {
            Thread.sleep(200);
            queueLock.lock();
            try {
                if (taskQueue.isEmpty()) break;
            } finally {
                queueLock.unlock();
            }
        }

        // Stop workers
        for (Thread t : threads) {
            t.interrupt();
        }
        for (Thread t : threads) {
            t.join();
        }

        // Write results to file
        try (BufferedWriter writer = new BufferedWriter(new FileWriter("output.txt"))) {
            for (String r : results) writer.write(r + "\n");
        } catch (IOException e) {
            logger.log(Level.SEVERE, "Error writing output.txt: " + e.getMessage(), e);
        }

        logger.info("All done.");
    }

    public static void main(String[] args) throws Exception {
        new Assignment6().run(5);
    }
}

File: assignment6.go
package main

import (
    "bufio"
    "fmt"
    "log"
    "math/rand"
    "os"
    "sync"
    "time"
)

type Task string
type Result string

func worker(id int, tasks <-chan Task, results chan<- Result, wg *sync.WaitGroup) {
    defer wg.Done()
    log.Printf("Worker-%d started.\n", id)
    for t := range tasks {
        func(task Task) {
            defer func() {
                if r := recover(); r != nil {
                    log.Printf("Worker-%d panic recovered: %v\n", id, r)
                }
            }()
            // simulate processing
            time.Sleep(time.Duration(500+rand.Intn(500)) * time.Millisecond)
            res := Result(fmt.Sprintf("Worker-%d processed: %s", id, task))
            results <- res
            log.Println(res)
        }(t)
    }
    log.Printf("Worker-%d exiting.\n", id)
}

func main() {
    rand.Seed(time.Now().UnixNano())
    logFile, err := os.OpenFile("rideshare_go.log", os.O_CREATE|os.O_APPEND|os.O_WRONLY, 0644)
    if err != nil {
        log.Fatalf("Cannot open log file: %v", err)
    }
    defer logFile.Close()
    log.SetOutput(logFile)

    taskChan := make(chan Task)
    resultChan := make(chan Result)
    var wg sync.WaitGroup

    numWorkers := 5
    wg.Add(numWorkers)
    for i := 1; i <= numWorkers; i++ {
        go worker(i, taskChan, resultChan, &wg)
    }

    // gather results
    go func() {
        f, err := os.Create("output_go.txt")
        if err != nil {
            log.Fatalf("Cannot create output file: %v", err)
        }
        defer f.Close()
        writer := bufio.NewWriter(f)
        for r := range resultChan {
            if _, err := writer.WriteString(string(r) + "\n"); err != nil {
                log.Printf("Error writing result: %v\n", err)
            }
        }
        writer.Flush()
    }()

    // enqueue tasks
    for i := 1; i <= 20; i++ {
        task := Task(fmt.Sprintf("RideRequest#%d", i))
        taskChan <- task
    }

    close(taskChan)  // signal workers no more tasks
    wg.Wait()
    close(resultChan) // signal result writer
    log.Println("Processing done.")
}